tools/harness-simulation/posix/sniffer_sim/sniffer.py

```diff
@@ -28,104 +28,172 @@
+
+        # Create transport
+        transport_factory = sniffer_transport.SnifferTransportFactory()
+        self._transport = transport_factory.create_transport()
+
+        # Start the sniffer main loop thread
         self._thread = threading.Thread(target=self._sniffer_main_loop)
         self._thread.daemon = True
-
         self._transport.open()
-
         self._thread_alive.set()
         self._thread.start()
 
-    def stop(self):
-        """ Stop sniffing. """
+        return sniffer_pb2.StatusReply(status=sniffer_pb2.OK)
 
-        self._thread_alive.clear()
+    def FilterNodes(self, request_iterator, context):
+        """ Only sniffer the specified nodes. """
 
-        self._transport.close()
+        self.logger.debug('call FilterNodes')
+
+        # Validate the state
+        if self._state != SnifferServicer.State.RUNNING:
+            return sniffer_pb2.StatusReply(status=sniffer_pb2.OPERATION_ERROR)
+
+        allowed_nodeids = set([x.nodeid for x in request_iterator])
+        # Validate the node IDs
+        for nodeid in allowed_nodeids:
+            if not 1 <= nodeid <= self.MAX_NODES_NUM:
+                return sniffer_pb2.StatusReply(status=sniffer_pb2.VALUE_ERROR)
+
+        self._mutex.acquire()
+        self._allowed_nodeids = allowed_nodeids
+        self._mutex.release()
+
+        return sniffer_pb2.StatusReply(status=sniffer_pb2.OK)
 
+    def Stop(self, request, context):
+        """ Stop sniffing, and return the pcap bytes. """
+
+        self.logger.debug('call Stop')
+
+        # Validate and change the state
+        if self._state != SnifferServicer.State.RUNNING:
+            return sniffer_pb2.StopReply(status=sniffer_pb2.OPERATION_ERROR, pcap_content=b'')
+        self._state = SnifferServicer.State.STOPPED
+
+        self._thread_alive.clear()
         self._thread.join(timeout=1)
-        self._thread = None
+        self._transport.close()
+
+        pcap_content = self._pcap.pop_all()
+        self._clear()
+
+        return sniffer_pb2.StopReply(status=sniffer_pb2.OK, pcap_content=pcap_content)
+
+
+def serve(address_port):
+    server = grpc.server(futures.ThreadPoolExecutor(max_workers=8))
+    sniffer_pb2_grpc.add_SnifferServicer_to_server(SnifferServicer(), server)
+    # add_secure_port requires a web domain
+    server.add_insecure_port(address_port)
+    logging.info('server starts on %s', address_port)
+    server.start()
 
-    def close(self):
-        """ Close the pcap file. """
+    def exit_handler(signum, context):
+        server.stop(1)
 
-        self._pcap.close()
+    signal.signal(signal.SIGINT, exit_handler)
+    signal.signal(signal.SIGTERM, exit_handler)
+
+    server.wait_for_termination()
 
 
 def run_sniffer():
+    logging.basicConfig(level=logging.INFO)
+
     parser = argparse.ArgumentParser()
-    parser.add_argument('-o',
-                        '--output',
-                        dest='output',
+    parser.add_argument('--address',
```

Comment:
When using `--address <ip> --port 50051`, it's not obvious what `address` and `port` means for the sniffer.

Is it clearer to use `--grpc-listen-addr <ip>:50051`or `--grpc-server <ip>:50051` instead?

Reply:
Now arguments use for format `--grpc-server <ip>:<port>`.